hashintel/hash
# Move `Agent` and dependent things to `stateful` crate

This is part of [Better separation of crates](https://app.asana.com/0/1201707629991362/1201481007343142/f) (_internal_).

## 🌟 What is the purpose of this PR?

This PR bundles together `Agent` related things (e.g. `AgentSchema`, `AgentStateField`, `Message`s) into the `stateful` crate (note, that this crates name may will be renamed later). The original goal was to move `AgentBatch` and `AgentPool` (thus the branch name), but those depend on `Agent` which is big enough to fill one PR.

## ⚠️ Known issues

- Converting `FieldSource` to a trait in #457 probably wasn't a good idea. this will be changed back in a follow-up PR to avoid merge conflicts
- Having the public module `agent` would trigger the (pedantic) [`stutter` clippy lint](https://rust-lang.github.io/rust-clippy/v0.0.212/index.html). This could be resolved by renaming `AgentSchema` to `Schema` and access it by `agent::Schema`, which would only be two more letters. Needs discussion and if desired should be changed in a follow-up PR.

## 🚫 Blocked by

- #457 

## 🐾 Next steps

- #471 
- (Rename `AgentSchema` (and similar things in `stateful::agent`) to `Schema` and access it by `agent::Schema`

## 🔍 What does this change?

- Make components to be moved independent:
    - [Add intra-doc links to proxy docs](https://github.com/hashintel/hash/pull/469/commits/99cc45fc8ebb7c18b78e1e724f2dabc65a522923)
    - [Make `PoolWriteProxy<AgentBatch>` functions free standing](https://github.com/hashintel/hash/pull/469/commits/f5f1d7a17944011b6b509e75f4f5e959b24fac39) (would be an implementation of a struct from an external crate)
    - [Make `PoolWriteProxy<MessageBatch>` functions free standing](https://github.com/hashintel/hash/pull/469/commits/2a8edf67d5266d0f2e4db06be704ad5cf0790d7b)
    - [Make `AgentSchema` generic over `FieldSource`](https://github.com/hashintel/hash/pull/469/commits/ecfc1e279e5f4d4a86433b686f3a9175f8063357)
    - [Make `IntoRecordBatch` generic over `FieldSource`](https://github.com/hashintel/hash/pull/469/commits/eec953b247d5572ea179d485b176a9ab028880f8)
- Move components:
    - [Move `BatchProxy` to `stateful::proxy`](https://github.com/hashintel/hash/pull/469/commits/8f92d1755b92ddeb67de2ecafb527b77ae74001e) (Required for `AgentBatch` and `MessageBatch` as that was the previous goal of this PR but it turned out that they depend on `Agent` which in terms had a bunch of other dependencies)
    - [Move `PoolProxy` to `stateful::proxy`](https://github.com/hashintel/hash/pull/469/commits/c89ecb2541138cd3bc8dbfdd34afc059dcf0e15c)
    - [Move `AgentSchema` to `stateful::agent`](https://github.com/hashintel/hash/pull/469/commits/6787d63e582aad036ac8b32f41e4c8a703a049c9)
    - [Move `MessageSchema` to `stateful::agent`](https://github.com/hashintel/hash/pull/469/commits/5878b99bc07802b8f66faf65042ca2be18c279ba)
    - [Move `AgentStateField` to `stateful::agent`](https://github.com/hashintel/hash/pull/469/commits/ad7bef4c2f189796d7ae6053d4d074829b53425d) (By this commit, `Agent`-related things are moved, but `Agent` also depends on `Outbound` messages)
    - [Move `GenericPayload` to `stateful::message`](https://github.com/hashintel/hash/pull/469/commits/cfda240827e5a7ea449f2ff8a25d4c82a5d705ec)
    - [Move `OutboundStopSimPayload` to `stateful::message`](https://github.com/hashintel/hash/pull/469/commits/bb22b440043dd0f090f3afcd4b832b33b68d59c5)
    - [Move `OutboundRemoveAgentPayload` to `stateful::message`](https://github.com/hashintel/hash/pull/469/commits/18d0c6e1e078d3155d6cd596153c9eb1ee6825b1) (By this commit, the independent payloads for `Outbound` are moved, `Agent` still requires some minor things like `Vec3`, `AgentName` (`Name` before) and `BUILTIN_FIELDS`)
    - [Move `BUILTIN_FIELDS` to `stateful::agent`](https://github.com/hashintel/hash/pull/469/commits/8bec253cebe79a38670fbb95bdec2f3714b12bec)
    - [Move `Vec3` to `stateful`](https://github.com/hashintel/hash/pull/469/commits/a8aec067c777cf7063034e5243ff3c7051cd9d47)
    - [Move `Name` to `stateful::agent::AgentName`](https://github.com/hashintel/hash/pull/469/commits/22d61e73f87808767eaf6f0159f5c765f8fe6c1c)
    - [Move `Agent` and `Outbound` to `stateful` 🎉 ](https://github.com/hashintel/hash/pull/469/commits/cb4b371c916bcf905726142b81b3da0cceedb6fc) (`Agent` and `Outbound` depends on each other, thus they are moved in one commit)

- Apply lints:
    - [Make RGB acronym lowercase](https://github.com/hashintel/hash/pull/469/commits/ee4c0e67f82ef211948f99e16e0ab21ffdef71fb)

- General structure changes with impact on `hash_engine_lib:
    - Access the message payloads by `payload::`
    - Access messages by `message::`

### 📜 Does this require a change to the docs?

The `agent` and `message` modules are not finished yet. I'll add module-level documentation once the rest of `Agent` things (like `AgentPool`) is merged into that crate. The same applies to `Message`s. We also want things like `Context` in this crate, so the crate level documentation will be adjusted once we are happy with the content of the crate. This is not an attempt to avoid writing documentation but to avoid duplicated work as the same documentation would need to be redesigned once that functionality is added.

The `proxy` module already had a documentation, which was slightly adjusted.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1202092280081544/f) _(internal)_

## ❓ How to test this?

Generally, this PR has very fine-grained commits and should be reviewed with this in mind. The big impact on line numbers mainly comes from moving symbols to another location. These commits don't introduce any logic changes.

- Verify the changes listed in *What does this change?* under *Make components to be moved independent*
- Verify the new structure shown below

## 📹 Demo

This results in the following structure:

```
stateful/
├─ agent/
│  ├─ Agent
│  ├─ AgentStateField
│  ├─ AgentName
│  ├─ AgentSchema
│  ├─ BUILTIN_FIELDS (to be made private)
├─ field/ (done in previous PR)
├─ message/
│  ├─ payload/
│  │  ├─ OutboundCreateAgent
│  │  ├─ OutboundRemoveAgent
│  │  ├─ OutboundRemoveAgentData
│  │  ├─ OutboundStopSim
│  │  ├─ OutboundStopSim
│  ├─ Outbound
│  ├─ CreateAgent
│  ├─ RemoveAgent
│  ├─ StopSim
│  ├─ MessageSchema
│  ├─ MESSAGE_ARROW_FIELDS (to be made private)
│  ├─ MESSAGE_BATCH_SCHEMA (to be made private)
│  ├─ MESSAGE_COLUMN_NAME (to be made private)
├─ proxy/
│  ├─ BatchReadProxy
│  ├─ BatchWriteProxy
│  ├─ PoolReadProxy
│  ├─ PoolWriteProxy
```
